vllm/attention/ops/flash_attn.py

```diff
@@ -0,0 +1,150 @@
+    # E.g., [0, 1, 2] means tokens are stored in 0th, 1st, and 2nd blocks
+    # in the kv cache. Each block can contain up to block_size tokens.
+    # 2nd dimensions are padded up to max_blocks_per_seq if it is cuda-graph
+    # captured.
+    block_tables: Optional[torch.Tensor]
+    kv_cache_dtype: str
+
+
+class FlashAttention:
+
+    @staticmethod
+    def get_supported_head_sizes() -> List[int]:
+        return [64, 80, 96, 112, 128, 256]
+
+    @staticmethod
+    def get_kv_cache_shape(
+        num_blocks: int,
+        block_size: int,
+        num_kv_heads: int,
+        head_size: int,
+    ) -> Tuple[int, ...]:
+        return (2, num_blocks, block_size * num_kv_heads * head_size)
+
+    @staticmethod
+    def split_kv_cache(
+        kv_cache: torch.Tensor,
+        num_kv_heads: int,
+        head_size: int,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        num_blocks = kv_cache.shape[1]
+
+        key_cache = kv_cache[0]
+        key_cache = key_cache.view(num_blocks, -1, num_kv_heads, head_size)
+
+        value_cache = kv_cache[1]
+        value_cache = value_cache.view(num_blocks, -1, num_kv_heads, head_size)
+        return key_cache, value_cache
+
+    @staticmethod
+    def write_to_paged_cache(
+        key: torch.Tensor,
+        value: torch.Tensor,
+        key_cache: torch.Tensor,
+        value_cache: torch.Tensor,
+        slot_mapping: torch.Tensor,
+        kv_cache_dtype: str,
+    ) -> None:
+        cache_ops.reshape_and_cache_flash(
+            key,
+            value,
+            key_cache,
+            value_cache,
+            slot_mapping.flatten(),
+            kv_cache_dtype,
+        )
+
+    @staticmethod
+    def forward_decode(
+        query: torch.Tensor,
+        key_cache: torch.Tensor,
+        value_cache: torch.Tensor,
+        block_tables: torch.Tensor,
+        context_lens: torch.Tensor,
+        max_context_len: int,
+        kv_cache_dtype: str,
+        num_kv_heads: int,
+        scale: float,
+        alibi_slopes: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        # TODO(skrider) tune num_splits heuristic
+        return flash_attn_with_kvcache(
+            query.unsqueeze(1),
+            key_cache,
+            value_cache,
+            None,
+            None,
+            cache_seqlens=context_lens,
+            block_table=block_tables,
+            softmax_scale=scale,
+            alibi_slopes=alibi_slopes,
+        )
+
+
+    @staticmethod
+    def forward_prefix(
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        key_cache: torch.Tensor,
+        value_cache: torch.Tensor,
+        block_tables: torch.Tensor,
+        subquery_start_loc: torch.Tensor,
+        prompt_lens_tensor: torch.Tensor,
+        context_lens: torch.Tensor,
+        max_subquery_len: int,
+        alibi_slopes: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        raise NotImplementedError
```

Comment:
@skrider Could you elaborate on why it is tricky to implement prefix-enabled attention?

Reply:
The flash API does not support attention to dense KV and paged KV in the same launch. We can either cache the dense KV before invoking forward_prefix or compute attention separately and merge with the online softmax trick. Not sure which would be best.

Reply:
@skrider isn't this possible with `flash_attn_varlen_func`? It seems like it supports paged kv cache (https://github.com/Dao-AILab/flash-attention/pull/831/files). 

IIUC, you can just pass k cache and v cache to k and v for `flash_attn_varlen_func` IIUC?

Reply:
Yes this is correct, however the kernel implements this by first copying k,v into the KV cache in gmem, and then reading the KV cache and computing attention, so we would need to ensure that there is enough space in the KV cache to append. And at that point why not just copy them in beforehand as is done with forward_decode, keeping the attention operator stateless as is implied.

Reply:
> We can either cache the dense KV before invoking forward_prefix

@skrider This is what our current Triton kernel is doing. Can't we do the same thing with flash-attn? Please take a look at the change I made in this PR.

Reply:
My mistake - we should be good then